plugin/action/throttle/throttle.go

```diff
@@ -102,40 +167,143 @@ func factory() (pipeline.AnyPlugin, pipeline.AnyConfig) {
 	return &Plugin{}, &Config{}
 }
 
+func (p *Plugin) syncWorker(ctx context.Context, jobCh <-chan limiter, wg *sync.WaitGroup) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-jobCh:
+			job.sync()
+			wg.Done()
+		}
+	}
+}
+
 func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.ActionPluginParams) {
 	p.config = config.(*Config)
+	p.logger = params.Logger
 	p.pipeline = params.PipelineName
-	p.limiterBuff = make([]byte, 0)
+	p.limiterBuf = make([]byte, 0)
+	ctx, cancel := context.WithCancel(context.Background())
+	p.ctx = ctx
+	p.cancel = cancel
 
 	limitersMu.Lock()
-	limiters[p.pipeline] = map[string]*limiter{}
+	limiters[p.pipeline] = map[string]limiter{}
 	limitersMu.Unlock()
 
-	for _, r := range p.config.Rules {
-		p.rules = append(p.rules, NewRule(r.Conditions, complexLimit{r.Limit, r.LimitKind}))
+	format, err := pipeline.ParseFormatName(p.config.TimeFieldFormat)
+	if err != nil {
+		format = p.format
+	}
+	p.format = format
+
+	if p.config.LimiterBackend == redisBackend {
+		if p.config.RedisBackendCfg.WorkerCount < 1 {
+			p.logger.Fatalf("workers_count must be > 0, passed: %d", p.config.RedisBackendCfg.WorkerCount)
+		}
+
+		p.redisClient = redis.NewClient(
+			&redis.Options{
+				Network:      "tcp",
+				Addr:         p.config.RedisBackendCfg.Host,
+				Password:     p.config.RedisBackendCfg.Password,
+				ReadTimeout:  p.config.RedisBackendCfg.Timeout_,
+				WriteTimeout: p.config.RedisBackendCfg.Timeout_,
+			})
+
+		if pingResp := p.redisClient.Ping(); pingResp.Err() != nil {
+			p.logger.Fatalf("can't ping redis: %s", pingResp.Err())
+		}
+
+		ticker := time.NewTicker(p.config.RedisBackendCfg.SyncInterval_)
+
+		limitersMu.Lock()
+		wg := sync.WaitGroup{}
+		// run syncer at single throttle for one pipeline
+		if _, ok := limiterSyncMap[p.pipeline]; !ok {
```

Comment:
It will not work, many goroutines may pass into this condition before `limiterSyncMap[p.pipeline] = struct{}{}`. Use `sync.Once()`

Reply:
They won't pass `limitersMu.Lock()` few lines above

Reply:
Actual flow: 
```
limiterMu.Lock()
if _, ok := limiterSyncMap[p.pipeline]; !ok {
     limiterSyncMap[p.pipeline] = struct{}{}
     limiterMu.Unlock()
     // init syncer ...
} else {
     limiterMu.Unlock()
}
```

It will' be like:

```
var (
    limiterMapOnce = map[string]sync.Once{} // It's empty
)

...

limiterMu.Lock() // anyway lock required to avoid race condition below
if _, ok := limiterMapOnce[p.pipeline]; !ok {
    syncForPipe := sync.Once{}
    limiterMapOnce[p.pipeline] = syncForPipe


    syncForPipe.Do(func() {  // yet this redundant, because next goroutines won't pass if _, ok := limiterMapOnce[p.pipeline]. So struct{}{} instead of sync.Once is enough
           // init syncer
    })
}


```